connect/clients/opensearch.py

```
@@ -0,0 +1,129 @@
+"""
+opensearch.py
+OpenSearch client wrapper
+"""
+import logging
+from connect.config import get_settings
+from opensearchpy import OpenSearch
+from opensearchpy.exceptions import RequestError
+from typing import Any, Optional
+
+
+logger = logging.getLogger(__name__)
+opensearch_client = None
+
+
+async def get_opensearch_client() -> Optional[OpenSearch]:
+    """
+    Create or return an OpenSearch client connected to the local
+    OpenSearch server.
+
+    :return: a connected OpenSearch client instance
+    """
+    global opensearch_client
+
+    if not opensearch_client:
+        opensearch_client = await create_opensearch_client()
+
+    return opensearch_client
+
+
+async def create_opensearch_client() -> Optional[OpenSearch]:
+    """
+    Create an OpenSearch client, connected to the configured OpenSearch server.
+
+    :return: a connected OpenSearch client instance
+    """
+    settings = get_settings()
+
+    # Create the client with SSL/TLS enabled, but hostname verification disabled
+    client = OpenSearch(
+        hosts=[{"host": settings.opensearch_server, "port": settings.opensearch_port}],
+        http_compress=True,  # enables gzip compression for request bodies
+        http_auth=(settings.opensearch_user, settings.opensearch_password),
+        use_ssl=True,
+        verify_certs=settings.certificate_verify,
+        ssl_assert_hostname=False,
+        ssl_show_warn=False,
+        ca_certs=settings.certificate_authority_path,
+    )
+
+    logger.info("Created OpenSearch client")
+    return client
+
+
+async def add_index(index: str):
+    """
+    Create a new index.
+    """
+    client = await get_opensearch_client()
+    try:
+        response = client.indices.create(index=index, body={})
```

Can we use a GET to check for index existence, rather than relying on an exception?
https://github.com/opensearch-project/opensearch-py/blob/main/opensearchpy/client/indices.py#L169

---

We can do a GET first, but in the case where the index doesn't exist yet, we would have to handle this exception:
opensearchpy.exceptions.NotFoundError: NotFoundError(404, 'index_not_found_exception', 'no such index [lpr]', lpr, index_or_alias)

We can do it either way, but I would opt for ignoring the exception on create if the index already exists (less code).

---

Got it. Thanks for clarifying that. Totally makes sense.